src/myrtlespeech/model/hard_lstm.py

```diff
@@ -0,0 +1,419 @@
+        self.input_size = input_size
+        self.hidden_size = hidden_size
+        self.num_layers = num_layers
+        self.num_directions = num_directions = 2 if bidirectional else 1
+        self.batch_first = batch_first
+        self.bias = bias
+        self.dropout = dropout
+        self.bidirectional = bidirectional
+
+        first_layer_args = [
+            input_size,
+            hidden_size,
+            forget_gate_bias,
+        ]
+        other_layer_args = [
+            hidden_size * num_directions,
+            hidden_size,
+            forget_gate_bias,
+        ]
+
+        layers = [layer_type(*first_layer_args)] + [
+            layer_type(*other_layer_args) for _ in range(num_layers - 1)
+        ]
+
+        self.layers = torch.nn.ModuleList(layers)
+
+    def forward(
+        self, input: Tensor, hx: Tuple[Tensor, Tensor]
+    ) -> Tuple[Tensor, Tuple[Tensor, Tensor]]:
+        """Performs forward pass of LSTM.
+
+        All inputs are moved to the GPU with :py:meth:`torch.nn.Module.cuda`
+        if :py:func:`torch.cuda.is_available` was :py:data:`True` on
+        initialisation.
+
+        Args:
+            input: A :py:class:`torch.Tensor` of rnn sequence inputs.
+
+            hx: The lstm hidden state: a Tuple of hidden state and cell state.
+
+        Returns:
+            A Tuple of Tuples; ``res = a, (b, c)`` where ``a`` is the lstm
+                output and ``(c, d)`` are the hidden and cell states of the
+                lstm.
+        """
+
+        if self.batch_first:
+            input = input.transpose(0, 1)
+
+        hn, cn = hx
+        req_size = self.num_layers, self.num_directions, hn.size(1), hn.size(2)
+        hn = hn.view(req_size)
+        cn = hn.view(req_size)
+
+        i = 0
+        output = input
+        hn_out = torch.jit.annotate(List[Tensor], [])
+        cn_out = torch.jit.annotate(List[Tensor], [])
+        for rnn_layer in self.layers:
+            output, (h, c) = rnn_layer(output, (hn[i], cn[i]))
+            hn_out += [h]
+            cn_out += [c]
+            i += 1
+
+        hn_out = torch.stack(hn_out)
+        cn_out = torch.stack(cn_out)
+        req_size_out = (-1, hn_out.size(2), hn_out.size(3))
+        out_states = hn_out.view(req_size_out), cn_out.view(req_size_out)
+
+        if self.batch_first:
+            output = output.transpose(0, 1)
+
+        return output, out_states
+
+
+class HardLSTMLayer(torch.nn.Module):
+    """A single-directional LSTM layer.
+
+    Args:
+        input_size: See :py:class:`HardLSTM`.
+
+        hidden_size: See :py:class:`HardLSTM`.
+
+        forget_gate_bias: See :py:class:`HardLSTM`.
+
+        reverse: A boolean. If True, this is is a reverse LSTM layer (used in
+            a bidirectional LSTM).
+    """
+
+    def __init__(
+        self,
+        input_size: int,
+        hidden_size: int,
+        forget_gate_bias: Optional[float] = None,
+        reverse: bool = False,
```

Comment:
`reverse` is not a property of a `HardLSTMLayer`, it only cares about iterating over some input in one direction. If a user/class wants this iteration to be done in the other direction (or some other permutation) it should be responsible for munging the input to that format. 

i.e. `reverse` is removed from `HardLSTMLayer` and `HardLSTMBidirLayer` just maintains two `HardLSTMLayer`s. In `HardLSTMBidirLayer.forward` one gets called with the standard input and the other gets called with the permuted/reversed input.

Reply:
Now outdated as reverse and forward layers are different classes